crates/katana/core/src/backend/gas_oracle.rs

```diff
@@ -1,25 +1,326 @@
+use std::backtrace::Backtrace;
+use std::collections::VecDeque;
+use std::fmt::Debug;
+use std::sync::Arc;
+
+use alloy_provider::{Provider, ProviderBuilder};
+use alloy_rpc_types_eth::BlockNumberOrTag;
+use alloy_transport::Transport;
+use anyhow::{Context, Ok};
 use katana_primitives::block::GasPrices;
+use katana_tasks::TaskSpawner;
+use tokio::sync::Mutex;
+use tokio::time::Duration;
+use url::Url;
+
+const BUFFER_SIZE: usize = 60;
+const INTERVAL: Duration = Duration::from_secs(60);
+const ONE_GWEI: u128 = 1_000_000_000;
 
 // TODO: implement a proper gas oracle function - sample the l1 gas and data gas prices
 // currently this just return the hardcoded value set from the cli or if not set, the default value.
 #[derive(Debug)]
-pub struct L1GasOracle {
+pub enum L1GasOracle {
+    Fixed(FixedL1GasOracle),
+    Sampled(SampledL1GasOracle),
+}
+
+#[derive(Debug)]
+pub struct FixedL1GasOracle {
+    gas_prices: GasPrices,
+    data_gas_prices: GasPrices,
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct SampledL1GasOracle {
+    prices: Arc<Mutex<SampledPrices>>,
+    l1_provider: Option<Url>,
+}
+
+#[derive(Debug, Default)]
+pub struct SampledPrices {
     gas_prices: GasPrices,
     data_gas_prices: GasPrices,
 }
 
+#[derive(Debug, Clone)]
+pub struct GasOracleWorker {
+    pub prices: Arc<Mutex<SampledPrices>>,
+    pub l1_provider_url: Url,
+    pub price_buffer: GasPriceBuffer,
+}
+
 impl L1GasOracle {
     pub fn fixed(gas_prices: GasPrices, data_gas_prices: GasPrices) -> Self {
-        Self { gas_prices, data_gas_prices }
+        L1GasOracle::Fixed(FixedL1GasOracle { gas_prices, data_gas_prices })
+    }
+
+    pub fn sampled(l1_provider: Option<Url>) -> Self {
```

Comment:
```suggestion
    pub fn sampled(l1_provider: Url) -> Self {
```

its an anti pattern to make this accept an `Option` only for it to be unwrapped down the line.